### PR TITLE
Credentials: Use QsortCompare instead of strcmp for type comparison

### PR DESCRIPTION
**Pull request by yashykt**

### CODE REVIEW

File: src/core/lib/security/credentials/credentials.h [hunk 1]
```diff
@@ -100,6 +100,8 @@ void grpc_override_well_known_credentials_path_getter(
 struct grpc_channel_credentials
     : grpc_core::RefCounted<grpc_channel_credentials> {
  public:
+  // \a type is used to downcast grpc_channel_credentials objects.
```
Comment (markdroth) [inline src/core/lib/security/credentials/credentials.h:103]: Let's be a little more explicit about the usage and requirements here:
- The *pointer value* is used to uniquely identify a creds implementation for down-casting purposes.
- Every creds impl should use a unique string instance, and it must use the same unique string instance for every creds instance of that creds impl.

File: src/core/lib/security/credentials/credentials.h [hunk 2]
```diff
@@ -185,6 +189,8 @@ using CredentialsMetadataArray = std::vector<std::pair<Slice, Slice>>;
 struct grpc_call_credentials
     : public grpc_core::RefCounted<grpc_call_credentials> {
  public:
+  // \a type is used to downcast grpc_call_credentials objects. Implementations
```
Comment (markdroth) [inline src/core/lib/security/credentials/credentials.h:192]: Same here.

File: src/core/lib/security/credentials/credentials.h (hunk 1 above)
Comment (yashykt) [reply src/core/lib/security/credentials/credentials.h:103]: > Every creds impl should use a unique string instance, and it must use the same unique string instance for every creds instance of that creds impl.

I think it's fine for an impl to use a different string if it wants to for some reason.

Comment (markdroth) [reply src/core/lib/security/credentials/credentials.h:103]: I can't think of a case where we'd actually want to do that.  Can you?

Comment (yashykt) [reply src/core/lib/security/credentials/credentials.h:103]: No, but I don't see a reason to enforce an additional requirement either

Comment (markdroth) [reply src/core/lib/security/credentials/credentials.h:103]: I'm not talking about enforcing a requirement, I'm talking about documenting the API contract.  The contract here is that the pointer address of the string will be used to deternine implementation equality.  That means that an implementation should always return the same pointer for all of its instances.

File: src/core/lib/security/credentials/credentials.h [hunk 3]
```diff
@@ -100,8 +100,10 @@ void grpc_override_well_known_credentials_path_getter(
 struct grpc_channel_credentials
     : grpc_core::RefCounted<grpc_channel_credentials> {
  public:
-  // \a type is used to downcast grpc_channel_credentials objects.
-  // Implementations should make sure to use unique types.
+  // The pointer value \a type is used to uniquely identify a creds
+  // implementation for down-casting purposes. Every creds implementation should
+  // use a unique string instance for every creds instance of that creds
```
Comment (markdroth) [inline src/core/lib/security/credentials/credentials.h:105]: The way this is currently phrased, it actually says the opposite of what we want here.  Saying "a unique string instance for every creds instance" means that each instance will use its own separate string.

I think what we want to say here is something like "Every creds implementation should use a unique string instance, which should be returned by all instances of that creds implementation."

File: src/core/lib/security/credentials/credentials.h [hunk 4]
```diff
@@ -189,8 +191,10 @@ using CredentialsMetadataArray = std::vector<std::pair<Slice, Slice>>;
 struct grpc_call_credentials
     : public grpc_core::RefCounted<grpc_call_credentials> {
  public:
-  // \a type is used to downcast grpc_call_credentials objects. Implementations
-  // should make sure to use unique types.
+  // The pointer value \a type is used to uniquely identify a creds
+  // implementation for down-casting purposes. Every creds implementation should
+  // use a unique string instance for every creds instance of that creds
```
Comment (markdroth) [inline src/core/lib/security/credentials/credentials.h:196]: Same here.

**Review by markdroth — APPROVED** [review]

**Pull request merged**